File: KysyNative/Sources/FocusLanguageEnforcer.swift
import Cocoa
import Combine

/// Watches which field is focused (system-wide) and, when it lands on a field
/// that must be typed in English (email / URL / password / payment), switches
/// the keyboard to an English input source. Runs in the background regardless of
/// whether the panel is open. Opt-in via Settings.
final class FocusLanguageEnforcer: ObservableObject {
    @Published var enabled: Bool = UserDefaults.standard.bool(forKey: "auto_lang_enabled") {
        didSet {
            UserDefaults.standard.set(enabled, forKey: "auto_lang_enabled")
            enabled ? start() : stop()
        }
    }
    @Published var lastAction: String?

    /// Field kinds (from FocusedField.guess) that should be typed in Latin.
    private let latinKinds: Set<String> = ["Email", "URL", "Password", "Payment"]

    private var timer: Timer?
    private let ownPid = getpid()
    private var lastSignature = ""

    func startIfEnabled() { if enabled { start() } }

    func start() {
        timer?.invalidate()
        // Light: only reads ~7 attributes per tick, and only acts on focus change.
        timer = Timer.scheduledTimer(withTimeInterval: 0.4, repeats: true) { [weak self] _ in
            self?.tick()
        }
    }

    func stop() {
        timer?.invalidate()
        timer = nil
        lastSignature = ""
    }

    private func tick() {
        guard AXIsProcessTrusted(),
              let field = AXInspector.focusedFieldLight(ignoring: ownPid) else { return }

        // Only act when focus actually moves to a new field.
        let sig = "\(field.pid ?? 0)|\(field.role ?? "")|\(field.placeholder ?? "")|\(field.title ?? "")"
        if sig == lastSignature { return }
        lastSignature = sig

        let kind = field.guess.label
        guard latinKinds.contains(kind) else { return }

        // Already on English? Nothing to do.
        guard let current = KeyboardLanguage.current(), !current.lang.hasPrefix("en") else { return }
        guard let english = KeyboardLanguage.firstEnglish() else { return }

        _ = KeyboardLanguage.select(id: english.id)
        DispatchQueue.main.async {
            self.lastAction = "Switched to \(english.name) for \(kind) field"
        }
    }
}

File: KysyWindows/Services/FocusLanguageEnforcer.cs
using System.Windows.Threading;

namespace Kysy.Services;

/// <summary>Watches the focused field and switches the keyboard to English when
/// it lands on an email / URL / password / payment field. Opt-in. Runs on a
/// light DispatcherTimer; acts only on focus change.</summary>
public sealed class FocusLanguageEnforcer
{
    private static readonly HashSet<string> LatinKinds = new() { "Email", "URL", "Password", "Payment" };
    private readonly DispatcherTimer _timer;
    private string _lastSignature = "";

    public bool Enabled
    {
        get => AppSettings.Current.AutoEnglishOnLatinFields;
        set
        {
            AppSettings.Current.AutoEnglishOnLatinFields = value;
            AppSettings.Save();
            if (value) _timer.Start(); else _timer.Stop();
        }
    }

    public string? LastAction { get; private set; }

    public FocusLanguageEnforcer()
    {
        _timer = new DispatcherTimer { Interval = TimeSpan.FromMilliseconds(400) };
        _timer.Tick += (_, _) => Tick();
    }

    public void StartIfEnabled() { if (Enabled) _timer.Start(); }

    private void Tick()
    {
        var field = FocusInspector.Focused();
        if (field is null) return;

        string sig = $"{field.ControlType}|{field.Name}|{field.IsPassword}";
        if (sig == _lastSignature) return;
        _lastSignature = sig;

        if (!LatinKinds.Contains(field.Guess)) return;

        // Already on a Latin layout (English, French, Spanish…)? Leave it — only
        // a non-Latin script (Hebrew, Arabic, Cyrillic…) garbles these fields.
        var current = LanguageManager.Current();
        if (current is null || LanguageManager.IsLatin(current.Hkl)) return;
        var target = LanguageManager.FirstLatin();
        if (target is null) return;

        LanguageManager.Switch(target.Hkl);
        LastAction = $"Switched to {target.Name} for {field.Guess} field";
    }
}
